File: samples/conference/registration-bc/src/main/java/com/microsoft/conference/registration/web/Controllers/PaymentController.cs
using System;
using System.Threading;
using System.Web.Mvc;
using ENode.Commanding;
using Payments.Commands;
using Registration.ReadModel;

namespace Registration.Web.Controllers
{
    public class PaymentController : Controller
    {
        private const int WaitTimeoutInSeconds = 5;

        private readonly ICommandService _commandService;
        private readonly IPaymentQueryService _paymentQueryService;

        public PaymentController(ICommandService commandService, IPaymentQueryService paymentQueryService)
        {
            this._commandService = commandService;
            this._paymentQueryService = paymentQueryService;
        }

        public ActionResult ThirdPartyProcessorPayment(string conferenceCode, Guid paymentId, string paymentAcceptedUrl, string paymentRejectedUrl)
        {
            var returnUrl = Url.Action("ThirdPartyProcessorPaymentAccepted", new { conferenceCode, paymentId, paymentAcceptedUrl });
            var cancelReturnUrl = Url.Action("ThirdPartyProcessorPaymentRejected", new { conferenceCode, paymentId, paymentRejectedUrl });

            var paymentDTO = this.WaitUntilAvailable(paymentId);
            if (paymentDTO == null)
            {
                return this.View("WaitForPayment");
            }

            var paymentProcessorUrl = this.Url.Action("Pay", "ThirdPartyProcessorPayment", new
            {
                area = "ThirdPartyProcessor",
                itemName = paymentDTO.Description,
                itemAmount = paymentDTO.TotalAmount,
                returnUrl,
                cancelReturnUrl
            });

            // redirect to external site
            return this.Redirect(paymentProcessorUrl);
        }

        public ActionResult ThirdPartyProcessorPaymentAccepted(string conferenceCode, Guid paymentId, string paymentAcceptedUrl)
        {
            this._commandService.SendAsync(new CompletePayment { AggregateRootId = paymentId });

            return this.Redirect(paymentAcceptedUrl);
        }

        public ActionResult ThirdPartyProcessorPaymentRejected(string conferenceCode, Guid paymentId, string paymentRejectedUrl)
        {
            this._commandService.SendAsync(new CancelPayment { AggregateRootId = paymentId });

            return this.Redirect(paymentRejectedUrl);
        }

        private Payment WaitUntilAvailable(Guid paymentId)
        {
            var deadline = DateTime.Now.AddSeconds(WaitTimeoutInSeconds);

            while (DateTime.Now < deadline)
            {
                var paymentDTO = this._paymentQueryService.FindPayment(paymentId);
                if (paymentDTO != null)
                {
                    return paymentDTO;
                }

                Thread.Sleep(500);
            }

            return null;
        }
    }
}

File: samples/conference/registration-bc/src/main/java/com/microsoft/conference/registration/controller/PaymentController.java
package com.microsoft.conference.registration.controller;


import com.microsoft.conference.common.ActionResult;
import com.microsoft.conference.common.payment.commands.CancelPayment;
import com.microsoft.conference.common.payment.commands.CompletePayment;
import com.microsoft.conference.registration.readmodel.service.PaymentQueryService;
import com.microsoft.conference.registration.readmodel.service.PaymentVO;
import lombok.var;
import org.enodeframework.commanding.ICommandService;
import org.enodeframework.common.io.Task;
import org.joda.time.DateTime;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.RestController;

@RestController
public class PaymentController {

    private int waitTimeoutInSeconds = 5;

    @Autowired
    private ICommandService commandService;

    @Autowired
    private PaymentQueryService paymentQueryService;

    public ActionResult<Boolean> thirdPartyProcessorPayment(String conferenceCode, String paymentId, String paymentAcceptedUrl, String paymentRejectedUrl) {
        PaymentVO paymentVO = this.waitUntilAvailable(paymentId);
        if (paymentVO == null) {
            return ActionResult.of(false);
        }
        return ActionResult.of(true);
    }

    public ActionResult<Void> thirdPartyProcessorPaymentAccepted(String conferenceCode, String paymentId, String paymentAcceptedUrl) {
        CompletePayment completePayment = new CompletePayment();
        completePayment.setAggregateRootId(paymentId);
        this.commandService.sendAsync(completePayment);
        return ActionResult.empty();
    }

    public ActionResult<Void> thirdPartyProcessorPaymentRejected(String conferenceCode, String paymentId, String paymentRejectedUrl) {
        CancelPayment cancelPayment = new CancelPayment();
        cancelPayment.setAggregateRootId(paymentId);
        this.commandService.sendAsync(cancelPayment);
        return ActionResult.empty();
    }

    private PaymentVO waitUntilAvailable(String paymentId) {
        DateTime future = new DateTime().plusSeconds(waitTimeoutInSeconds);
        while (future.isAfterNow()) {
            var paymentDTO = this.paymentQueryService.findPayment(paymentId);
            if (paymentDTO != null) {
                return paymentDTO;
            }
            Task.sleep(500);
        }
        return null;
    }
}
